### PR TITLE
[Optimize](invert index) Optimize multiple terms conjunction query

### PR DESCRIPTION
**Pull request by zzzxl1993**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/olap/rowset/segment_v2/inverted_index_reader.cpp [hunk 1]
```diff
@@ -436,6 +386,68 @@ Status FullTextIndexReader::query(OlapReaderStatistics* stats, const std::string
...
+        const IndexSearcherPtr& index_searcher, bool& null_bitmap_already_read,
+        const std::unique_ptr<lucene::search::Query>& query,
+        const std::shared_ptr<roaring::Roaring>& term_match_bitmap) {
+    check_null_bitmap(index_searcher, null_bitmap_already_read);
+
+    try {
+        SCOPED_RAW_TIMER(&stats->inverted_index_searcher_search_timer);
+        if (query_type == InvertedIndexQueryType::MATCH_ANY_QUERY ||
+            query_type == InvertedIndexQueryType::EQUAL_QUERY) {
+            index_searcher->_search(query.get(), [&term_match_bitmap](DocRange* docRange) {
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/inverted_index_reader.cpp:400]: use doc_range consistently

File: be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            _others.push_back(iterators[i]);
+        }
+    }
+
+    if (iterators.size() >= 2) {
+        int32_t little = iterators[0].docFreq();
+        int32_t big = iterators[iterators.size() - 1].docFreq();
+        if (little == 0) {
+            _useSkip = true;
+        } else if ((big / little) > 1000) {
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp:54]: make it a session variable

**Review by airborne12 — APPROVED** [review]
LGTM

**Comment by zzzxl1993** [comment]
run build all

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

File: be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+            _CLDELETE(term);
+        }
+    }
+    for (auto& termDoc : _termDocs) {
+        if (termDoc) {
+            _CLDELETE(termDoc);
+        }
+    }
+}
+
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp:37]: combine to 'little == 0 || (big / little) > _conjunction_ratio'

File: be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.h [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        _conjunction_ratio = conjunction_ratio;
+    }
+
+    void add(const std::wstring& fieldName, const std::vector<std::wstring>& wterms);
+    void search(roaring::Roaring& roaring);
+
+private:
+    void search_by_bitmap(roaring::Roaring& roaring);
+
+    void search_by_skiplist(roaring::Roaring& roaring) {
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.h:45]: it's confusing to put search_by_skiplist and next_doc impl in .h

File: be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.h [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        int32_t doc = 0;
+        while ((doc = next_doc()) != INT32_MAX) {
+            roaring.add(doc);
+        }
+    }
+
+    int32_t next_doc() { return do_next(_lead1.nextDoc()); }
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.h:52]: next_doc is only used once and is simple, is it necessary to create a new function for it?

File: be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    search_by_skiplist(roaring);
+}
+
+void ConjunctionQuery::search_by_bitmap(roaring::Roaring& roaring) {
+    auto func = [&roaring](const TermIterator& termDocs) {
+        roaring::Roaring result;
+        DocRange docRange;
+        while (termDocs.readRange(&docRange)) {
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp:97]: add comment for this algrithm

File: be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+            if (docRange.type_ == DocRangeType::kMany) {
+                result.addMany(docRange.doc_many_size_, docRange.doc_many->data());
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp:99]: while (true)

File: be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+            if (docRange.type_ == DocRangeType::kMany) {
+                roaring.addMany(docRange.doc_many_size_, docRange.doc_many->data());
+            } else {
+                roaring.addRange(docRange.doc_range.first, docRange.doc_range.second);
+            }
+        }
+    }
+
+    if (!_lead2.isEmpty()) {
+        func(_lead2);
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp:119]: goto is not suggesed to be used.

File: be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.h [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+    int32_t do_next(int32_t doc);
+
+private:
+    IndexReader* _reader = nullptr;
+    IndexVersion _indexVersion = IndexVersion::kV0;
+    int32_t _conjunction_ratio = 1000;
+    bool _useSkip = false;
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.h:60]: _use_skiplist

File: be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+
+void ConjunctionQuery::add(const std::wstring& fieldName, const std::vector<std::wstring>& wterms) {
+    if (wterms.size() < 1) {
+        _CLTHROWA(CL_ERR_IllegalArgument, "ConjunctionQuery::add: terms.size() < 1");
+    }
+
+    std::vector<TermIterator> iterators;
+    for (auto& wterm : wterms) {
+        Term* t = _CLNEW Term(fieldName.c_str(), wterm.c_str());
+        _terms.push_back(t);
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp:46]: can be merged to set _use_skiplist and just select search_by_bitmap/search_by_skiplist by _use_skiplist

**Comment by zzzxl1993** [comment]
run p0

File: be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+        for (auto& other : _others) {
+            if (other.isEmpty()) {
+                continue;
+            }
+
+            if (other.docID() < doc) {
+                int32_t next = other.advance(doc);
+                if (next > doc) {
+                    doc = _lead1.advance(next);
+                    goto advanceHead;
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp:153]: goto is not suggesed to be used.

File: be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+        if (next2 != doc) {
+            doc = _lead1.advance(next2);
+            if (next2 != doc) {
+                continue;
+            }
+        }
+
+        for (auto& other : _others) {
+            if (other.isEmpty()) {
+                continue;
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp:146]: continue others for loop or outer for loop?

File: be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+
+    if (!_useSkip) {
+        search_by_bitmap(roaring);
+        return;
+    }
+
+    search_by_skiplist(roaring);
+}
+
+void ConjunctionQuery::search_by_bitmap(roaring::Roaring& roaring) {
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp:93]: add comment for this algrithm

File: be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+    for (auto& other : _others) {
+        if (other.isEmpty()) {
+            continue;
+        }
+
+        func(other);
+    }
+}
+
+int32_t ConjunctionQuery::do_next(int32_t doc) {
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp:131]: add comment for this algrithm

File: be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+        } else if ((big / little) > _conjunction_ratio) {
+            _useSkip = true;
+        }
+    }
+}
+
+void ConjunctionQuery::search(roaring::Roaring& roaring) {
+    if (_lead1.isEmpty()) return;
+
+    if (_indexVersion == IndexVersion::kV0) {
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp:80]: can be merged to set _use_skiplist and just select search_by_bitmap/search_by_skiplist by _use_skiplist

File: be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+        _lead2 = iterators[1];
+        for (int32_t i = 2; i < _terms.size(); i++) {
+            _others.push_back(iterators[i]);
+        }
+    }
+
+    if (iterators.size() >= 2) {
+        int32_t little = iterators[0].docFreq();
+        int32_t big = iterators[iterators.size() - 1].docFreq();
+        if (little == 0) {
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp:69]: combine to 'little == 0 || (big / little) > _conjunction_ratio'

**Review by xiaokang — DISMISSED** [review]
LGTM

**Comment by xiaokang** [comment]
run buildall

**Comment by shuke987** [comment]
run clickbench

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run fe

**Comment by zzzxl1993** [comment]
run beut

**Comment by zzzxl1993** [comment]
run buildall

**Review by xiaokang — APPROVED** [review]
LGTM

**Pull request merged**